lib/hashie/utils.rb

```diff
@@ -12,5 +12,33 @@ def self.method_information(bound_method)
         "defined in #{bound_method.owner}"
       end
     end
+
+    # Duplicates a value or returns the value when it is not duplicable
+    #
+    # @api public
+    #
+    # @param value [Object] the value to safely duplicate
+    # @return [Object] the duplicated value
+    def self.safe_dup(value)
+      case value
+      when Complex, FalseClass, NilClass, Rational, Method, Symbol, TrueClass, *integer_classes
+        value
+      else
+        value.dup
+      end
+    end
```

Comment:
Is this a pattern from other libraries? I feel like taking extra overhead to call `dup` on even those simple types is OK because you can't make assumptions about any underlying implementation. It's also unclear whether there's performance gain doing a `case` vs. a `dup`.

Reply:
The issue is that on Rubies < 2.4(? I think ... maybe 2.5), these classes raise a `TypeError` when you call `#dup` on them. Since they are all written in C, they need to have an `allocate` method for `#dup` to work and they do not.

I can roll out the safe dup change and run it on CI if you'd like to see. Otherwise, you can pull it down and try running the test suite after changing the `Hashie::Utils.safe_dup(value)` line to `value.dup`.

I originally went the route of a refinement that added [ActiveSupport-like feature detection](https://github.com/rails/rails/blob/master/activesupport/lib/active_support/core_ext/object/duplicable.rb) for `#dup, but decided that was a lot of overhead to do something that is just as fast and much less code.

What do you think?

Reply:
Ok, got it. I think what you did makes sense then, I didn't know this.